File: com.trollworks.gcs/src/com/trollworks/gcs/prereq/SkillPrereq.java
/*
 * Copyright ©1998-2021 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package com.trollworks.gcs.prereq;

import com.trollworks.gcs.character.GURPSCharacter;
import com.trollworks.gcs.criteria.IntegerCriteria;
import com.trollworks.gcs.criteria.StringCompareType;
import com.trollworks.gcs.criteria.StringCriteria;
import com.trollworks.gcs.datafile.LoadState;
import com.trollworks.gcs.skill.Skill;
import com.trollworks.gcs.ui.widget.outline.ListRow;
import com.trollworks.gcs.utility.I18n;
import com.trollworks.gcs.utility.json.JsonMap;
import com.trollworks.gcs.utility.json.JsonWriter;

import java.io.IOException;
import java.text.MessageFormat;
import java.util.Map;
import java.util.Set;

/** A Skill prerequisite. */
public class SkillPrereq extends NameLevelPrereq {
    /** The XML tag for this class. */
    public static final  String         TAG_ROOT           = "skill_prereq";
    private static final String         TAG_SPECIALIZATION = "specialization";
    private              StringCriteria mSpecializationCriteria;

    /**
     * Creates a new prerequisite.
     *
     * @param parent The owning prerequisite list, if any.
     */
    public SkillPrereq(PrereqList parent) {
        super(TAG_ROOT, parent);
        mSpecializationCriteria = new StringCriteria(StringCompareType.ANY, "");
    }

    /**
     * Loads a prerequisite.
     *
     * @param parent The owning prerequisite list, if any.
     * @param m      The {@link JsonMap} to load from.
     */
    public SkillPrereq(PrereqList parent, JsonMap m) throws IOException {
        super(parent, m);
    }

    private SkillPrereq(PrereqList parent, SkillPrereq prereq) {
        super(parent, prereq);
        mSpecializationCriteria = new StringCriteria(prereq.mSpecializationCriteria);
    }

    @Override
    protected void initializeForLoad() {
        mSpecializationCriteria = new StringCriteria(StringCompareType.ANY, "");
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) {
            return true;
        }
        if (obj instanceof SkillPrereq && super.equals(obj)) {
            return mSpecializationCriteria.equals(((SkillPrereq) obj).mSpecializationCriteria);
        }
        return false;
    }

    @Override
    public void loadSelf(JsonMap m, LoadState state) throws IOException {
        super.loadSelf(m, state);
        mSpecializationCriteria.load(m.getMap(TAG_SPECIALIZATION));
    }

    @Override
    public void saveSelf(JsonWriter w) throws IOException {
        super.saveSelf(w);
        mSpecializationCriteria.save(w, TAG_SPECIALIZATION);
    }

    @Override
    public String getJSONTypeName() {
        return TAG_ROOT;
    }

    @Override
    public Prereq clone(PrereqList parent) {
        return new SkillPrereq(parent, this);
    }

    @Override
    public boolean satisfied(GURPSCharacter character, ListRow exclude, StringBuilder builder, String prefix) {
        boolean         satisfied     = false;
        String          techLevel     = null;
        StringCriteria  nameCriteria  = getNameCriteria();
        IntegerCriteria levelCriteria = getLevelCriteria();

        if (exclude instanceof Skill) {
            techLevel = ((Skill) exclude).getTechLevel();
        }

        for (Skill skill : character.getSkillsIterator()) {
            if (exclude != skill && nameCriteria.matches(skill.getName()) && mSpecializationCriteria.matches(skill.getSpecialization())) {
                satisfied = levelCriteria.matches(skill.getLevel());
                if (satisfied && techLevel != null) {
                    String otherTL = skill.getTechLevel();
                    satisfied = otherTL == null || techLevel.equals(otherTL);
                }
                if (satisfied) {
                    break;
                }
            }
        }
        if (!has()) {
            satisfied = !satisfied;
        }
        if (!satisfied && builder != null) {
            builder.append(MessageFormat.format(I18n.Text("{0}{1} a skill whose name {2}"), prefix, hasText(), nameCriteria.toString()));
            boolean notAnySpecialization = !mSpecializationCriteria.isTypeAnything();
            if (notAnySpecialization) {
                builder.append(MessageFormat.format(I18n.Text(", specialization {0},"), mSpecializationCriteria.toString()));
            }
            if (techLevel == null) {
                builder.append(MessageFormat.format(I18n.Text(" and level {0}"), levelCriteria.toString()));
            } else {
                if (notAnySpecialization) {
                    builder.append(",");
                }
                builder.append(MessageFormat.format(I18n.Text(" level {0} and tech level matches\n"), levelCriteria.toString()));
            }
        }
        return satisfied;
    }

    @Override
    public void fillWithNameableKeys(Set<String> set) {
        super.fillWithNameableKeys(set);
        ListRow.extractNameables(set, mSpecializationCriteria.getQualifier());
    }

    @Override
    public void applyNameableKeys(Map<String, String> map) {
        super.applyNameableKeys(map);
        mSpecializationCriteria.setQualifier(ListRow.nameNameables(map, mSpecializationCriteria.getQualifier()));
    }

    /** @return The specialization comparison object. */
    public StringCriteria getSpecializationCriteria() {
        return mSpecializationCriteria;
    }
}

File: model/skill_prereq.go
/*
 * Copyright ©1998-2022 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package model

import (
	"github.com/richardwilkes/gcs/v5/model/criteria"
	"github.com/richardwilkes/toolbox/i18n"
	"github.com/richardwilkes/toolbox/xio"
)

var _ Prereq = &SkillPrereq{}

// SkillPrereq holds a prerequisite for a skill.
type SkillPrereq struct {
	Parent                 *PrereqList      `json:"-"`
	Type                   PrereqType       `json:"type"`
	Has                    bool             `json:"has"`
	NameCriteria           criteria.String  `json:"name,omitempty"`
	LevelCriteria          criteria.Numeric `json:"level,omitempty"`
	SpecializationCriteria criteria.String  `json:"specialization,omitempty"`
}

// NewSkillPrereq creates a new SkillPrereq.
func NewSkillPrereq() *SkillPrereq {
	return &SkillPrereq{
		Type: SkillPrereqType,
		NameCriteria: criteria.String{
			StringData: criteria.StringData{
				Compare: criteria.Is,
			},
		},
		LevelCriteria: criteria.Numeric{
			NumericData: criteria.NumericData{
				Compare: criteria.AtLeast,
			},
		},
		SpecializationCriteria: criteria.String{
			StringData: criteria.StringData{
				Compare: criteria.Any,
			},
		},
		Has: true,
	}
}

// PrereqType implements Prereq.
func (s *SkillPrereq) PrereqType() PrereqType {
	return s.Type
}

// ParentList implements Prereq.
func (s *SkillPrereq) ParentList() *PrereqList {
	return s.Parent
}

// Clone implements Prereq.
func (s *SkillPrereq) Clone(parent *PrereqList) Prereq {
	clone := *s
	clone.Parent = parent
	return &clone
}

// FillWithNameableKeys implements Prereq.
func (s *SkillPrereq) FillWithNameableKeys(m map[string]string) {
	Extract(s.NameCriteria.Qualifier, m)
	Extract(s.SpecializationCriteria.Qualifier, m)
}

// ApplyNameableKeys implements Prereq.
func (s *SkillPrereq) ApplyNameableKeys(m map[string]string) {
	s.NameCriteria.Qualifier = Apply(s.NameCriteria.Qualifier, m)
	s.SpecializationCriteria.Qualifier = Apply(s.SpecializationCriteria.Qualifier, m)
}

// Satisfied implements Prereq.
func (s *SkillPrereq) Satisfied(entity *Entity, exclude any, tooltip *xio.ByteBuffer, prefix string, _ *bool) bool {
	satisfied := false
	var techLevel *string
	if sk, ok := exclude.(*Skill); ok {
		techLevel = sk.TechLevel
	}
	Traverse(func(sk *Skill) bool {
		if exclude == sk || !s.NameCriteria.Matches(sk.Name) || !s.SpecializationCriteria.Matches(sk.Specialization) {
			return false
		}
		satisfied = s.LevelCriteria.Matches(sk.LevelData.Level)
		if satisfied && techLevel != nil {
			satisfied = sk.TechLevel == nil || *techLevel == *sk.TechLevel
		}
		return satisfied
	}, false, true, entity.Skills...)
	if !s.Has {
		satisfied = !satisfied
	}
	if !satisfied && tooltip != nil {
		tooltip.WriteString(prefix)
		tooltip.WriteString(HasText(s.Has))
		tooltip.WriteString(i18n.Text(" a skill whose name "))
		tooltip.WriteString(s.NameCriteria.String())
		if s.SpecializationCriteria.Compare != criteria.Any {
			tooltip.WriteString(i18n.Text(", specialization "))
			tooltip.WriteString(s.SpecializationCriteria.String())
			tooltip.WriteByte(',')
		}
		if techLevel == nil {
			tooltip.WriteString(i18n.Text(" and level "))
			tooltip.WriteString(s.LevelCriteria.String())
		} else {
			if s.SpecializationCriteria.Compare != criteria.Any {
				tooltip.WriteByte(',')
			}
			tooltip.WriteString(i18n.Text(" level "))
			tooltip.WriteString(s.LevelCriteria.String())
			tooltip.WriteString(i18n.Text(" and tech level matches"))
		}
	}
	return satisfied
}
